### DISCUSSION TITLE
set strategy for volume vacuum

### DISCUSSION THREAD

**Discussion by guol-fnst** [seaweedfs/seaweedfs#3349] [Ideas]
Currently, weedfs start to vacuum when  garbage percent is over threshold  immediatly,no matter how busy the cluster is. Besides when vacuuming, read or write operation for this volume will be suspened.

So we can set strategy  for volume vacuum to reduce cluster load and unavailable time.
For example, we can start  vacuuming when the following condition  happens:
1) only  vacuuming on  weekends
2) cpu usage is bellow xx
3) if this volume has not been  read or writeen for certain time

**Comment by kmlebedev** [comment]
@guol-fnst I don't understand what exactly is the pain?

Cleaning is done in the background and doesn’t interfere, for example, I clean in the background if more than 30% and on weekends crone job if more than 5%
```
kubectl exec -i fast-master-0 -- sh -c 'echo -e "lock\nvolume.vacuum -garbageThreshold=0.05\n" | weed shell'
```

**Reply by guol-fnst** [reply]
For example. When system is on heavy load, and  garbage percent of the volume is over 30, at the same time the volume is  read or written frequently. If seaweedfs start to vacuum this volume, then read or write may be suspend.

**Reply by guol-fnst** [reply]
This volume will not be written on  CommitCompact
https://github.com/chrislusf/seaweedfs/blob/7a6c559ab4a6b696bb574454b297ebefabec29ed/weed/storage/volume_vacuum.go#L105

**Reply by kmlebedev** [reply]
I don't see a problem with increasing the number of writeable volumes
for reading i use replica

 heavy load can be checked through
```
test $(echo "$(grep -oP '^\d+\.\d+' /proc/loadavg) * 100"|bc) -lt 100 && echo "do vacuum"
```

**Reply by guol-fnst** [reply]
In our env ,we use raid instead of replica. 
When cacuuming volume, filer can read from replica volume. But writing may be stuck.
I encourtered write timeout several times. I guess it's the reason.

**Reply by guol-fnst** [reply]
Also thanks for the advice of getting cpu useage. Our system has cpu/memory monitor  process ,this can be get easily. I just want to make seaweedfs more flexable.

**Reply by kmlebedev** [reply]
It’s difficult not to cost too much, this task pulls on a separate subsystem of the task scheduler
Since there is also volume.balance

I think you initially chose the wrong design without replicas

**Reply by guol-fnst** [reply]
I get your point! In some situation replica is not necessary  and it wastes  disk.
Now that this stuck may happen. I have a new design to  reduce commiting time through optimizing CommitCompact() .  If this design works , this proposal will be useless.
Thanks again!

**Reply by kmlebedev** [reply]
> I get your point! In some situation replica is not necessary and it wastes disk. I have a new design to reduce commiting time through optimizing CommitCompact() . If this design works , this proposal will be useless. Thanks again!

A replica is needed, since the north and the whole data centers are not immortal and sometimes they need to be decommissioned for maintenance (updating firmware, drivers, OS and power and communication systems), how do you plan to do this?
Or are your servers immortal?

**Reply by guol-fnst** [reply]
We have independent storage cluster which provides replicated shared folder for weed volumes.
If a weed volume server is crashed. We just start a new volume server on other server and dont need to do compicated recovery work. This can be done using  script.

**Comment by kmlebedev** [comment]
Now it hurts me too https://github.com/chrislusf/seaweedfs/issues/3369
